### DISCUSSION TITLE
Best way to prevent a Man-in-the-middle attack in iOS

### DISCUSSION THREAD

**Discussion by dovber-persiko-healthy** [MobileNativeFoundation/discussions#51] [General]
I'm trying to figure out the best way to prevent a Man-in-the-middle attack.

I see a lot of information about SSL Pinning,
But there are a lot of downsides to the pinning solutions.

I see another solution called Certificate Transparency (CT), Which is most common in browsers.

I made a test project and CT protocol using NSRequiresCertificateTransparency works as expected without any downsides, 
But maybe I am missing something.

I'm wondering what is the best solution in iOS SSL Pinning or CT?

Thanks

**Comment by dalemyers** [comment]
We use pinned certificates in Outlook. However, it's always worth considering what you are trying to prevent/protect. If someone has a jail broken device, it's relatively simple for them to modify the binary to bypass the pinning check and see the traffic.

**Reply by donnie-jp** [reply]
Nice to see pinning being used.

Are you pinning against the cert or the public key info?
Are you embedding it in your app?
How do you mitigate against the backend cert/key being changed?

**Reply by dalemyers** [reply]
@icodesign knows the details of this.

**Reply by kumarc-ps** [reply]
SSL Pinning is the solution, we also prefer for our enterprise app. If you're using Alamofire for ReSTful APIs it supports session configuration with pinned certificates.

**Reply by icodesign** [reply]
Normally you just need check against the root certificates (we embed some well known root certs in our app). Previous builds won't work if your root cert changes but it's not embedded before (less likely to happen regularly). But anyway, you can prepare ahead, right?

**Reply by donnie-jp** [reply]
@icodesign Thanks for the info.

Some orgs have been caught out by the embedding aspect and not prepared ahead.

Are the certs you're embedding from [outlook/microsoft] servers that you control? Or are they further up the chain?

I got this feedback from our security team's mobile auditor about pinning:

> Ideally you should be pinning the public key of the final leaf cert - this prevents a hijack in the middle of the chain from affecting it, but can make rotation and revocation more complex. I think the preferred way to deal with that is to pin based on the modulus of the pubkey, because they you can change out the cert without a problem so long as the private key does not change. You can then also do a hot/cold privatekey and store both moduli in the app to let you swap to a newer one later without an outage. This is effective at preventing MitM.

And on the management of certs/keys for mobile apps:

> ... It adds complexity and management overhead. You now have to private keys to protect. Its tempting to have long key lifetimes, so you need to change it less often - but that increases the risk of ‘not remembering how to renew/fix it’ next time, and increases the window of time a stolen private key could be useful for. Solid documentation, including procedures on how to rotate those certs is key to avoiding that, as is tracking access to, and usage of, that private key.

I'm not a mobile security specialist but that advice seems solid. 
It also chimes with the [pinning cheatsheet from OWASP](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Pinning_Cheat_Sheet.md#ios).

**Reply by Exadra37** [reply]
> I got this feedback from our security team's mobile auditor about pinning:
>
>    Ideally you should be pinning the public key of the final leaf cert - this prevents a hijack in the middle of the chain from affecting it, but can make rotation and revocation more complex. I think the preferred way to deal with that is to pin based on the modulus of the pubkey, because they you can change out the cert without a problem so long as the private key does not change. You can then also do a hot/cold privatekey and store both moduli in the app to let you swap to a newer one later without an outage. This is effective at preventing MitM.

I work as a [Developer Advocate](https://stackoverflow.com/users/6454622/exadra37?tab=profile) for security in APIs and Mobile apps and I can confirm that pining the leaf certificate as @donnie-jp mentions is the preferred way of doing it. 

Anyway, bear in mind that pinning can be bypassed during runtime with instrumentation frameworks, like Frida, or by repackaging the mobile app. To make it hard to bypass a mobile app also needs to have runtime self defense mechanisms to detect this situations. 

> You can then also do a hot/cold privatekey and store both moduli in the app to let you swap to a newer one later without an outage.

A word of caution for how you keep private keys secure. The hot private key(the one being used currently in production) must be physically stored in a different physical space from the cold private key(the backup one). This is necessary to avoid the compromise of both private keys simultaneously and for the cases where a catastrophe strikes.

>     ... It adds complexity and management overhead. You now have to private keys to protect. Its tempting to have long key lifetimes, so you need to change it less often - but that increases the risk of ‘not remembering how to renew/fix it’ next time, and increases the window of time a stolen private key could be useful for. Solid documentation, including procedures on how to rotate those certs is key to avoiding that, as is tracking access to, and usage of, that private key.

At work I have an [Android mobile app demo](https://github.com/approov/shipfast-api-protection) running with an API backend that uses LetsEncrypt certificates that the mobile app pins against the public key, and  as all of you may know LetsEncrypt certificates last three months, but once the pinning is done against the public key the new certificate will keep the same pin, therefore the mobile app will continue to operate as usual. 

While the mobile app is for Android what matters here is the fact of pinning being done against the public key of the leaf certificate, instead of the embedded certificates in the mobile app, that add operational complexity, and as said above in the quote, it requires solid procedures and documentation to be in place to avoid to shoot yourself in the foot.

**Reply by sherlock-holmes-221** [reply]
What needs to be done if your app is talking to a 3rd party server as well? For example - Firebase Analytics.

**Reply by Exadra37** [reply]
Pinning against Third Party APIs is not advised, because you don't have control from when they will rotate their certificates, but if you do it you need to have a secure way of doing over the air updates to your mobile app instances in order to change the pins it is matching against.

I wrote the article [Using a Reverse Proxy to Protect Third Party APIs](https://approov.io/blog/using-a-reverse-proxy-to-protect-third-party-apis) that goes more in detail about this type of situation:

> In this article you will start by learning what Third Party APIs are, and why you shouldn’t access them directly from within your mobile app. Next you will learn what a Reverse Proxy is, followed by when and why you should use it to protect the access to the Third Party APIs used in your mobile app.
> A recurring theme in this article was the advice not to access Third Party APIs directly from a mobile app. As we have discussed, once your mobile app is released any secret in it becomes public, thus up for grabs by attackers to use on your behalf. If you are not careful you will be the one paying the bill or finding that your free tier resources have been exhausted by someone else.

**Comment by FranzBusch** [comment]
We use [Approov](https://www.approov.io) at Sixt to protect some of our endpoints against Man-in-the-middle attacks but also malicious actors with Jailbroken devices.

Approov supports a bunch of different security mechanisms but an important one from them is dynamic SSL pinning which prevents most of the MITM attacks.

**Reply by Exadra37** [reply]
> Approov supports a bunch of different security mechanisms but an important one from them is dynamic SSL pinning which prevents most of the MITM attacks.

Dynamic certificate pinning means that you don't need to release a new mobile app to your clients in order to rotate the certificates pins, because they can be securely updated via over the air updates, aka dynamic configuration, and when someone tries to bypass pinning it will be detected and then the API can refuse the request from the mobile app being tampered with.

> Disclaimer: I am a Developer Advocate for Mobile API Security at Approov.
